src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataReaderExtensions.cs

```diff
@@ -171,11 +173,58 @@ private static void ReadCollectionItem(ODataReader reader, Stack<ODataItemBase>
                     }
 
                     break;
+                case ODataReaderState.DeltaResourceSetStart:
+                    ODataDeltaResourceSet deltaResourceSet = (ODataDeltaResourceSet)reader.Item;
+                    Contract.Assert(deltaResourceSet != null, "ResourceSet should never be null.");
+
+                    ODataDeltaResourceSetWrapper deltaResourceSetWrapper = new ODataDeltaResourceSetWrapper(deltaResourceSet);
+
+                    if (itemsStack.Count > 0)
+                    {
+                        AddNestedResourceInfo(itemsStack, deltaResourceSetWrapper);
+                    }
+                    else
+                    {
+                        topLevelItem = deltaResourceSetWrapper;
+                    }
+
+                    itemsStack.Push(deltaResourceSetWrapper);
+                    break;
+
+                case ODataReaderState.DeltaLink:
+                case ODataReaderState.DeltaDeletedLink:
+
+                    //Throw error if Delta Link appears
+                    throw Error.NotSupported(SRResources.DeltaLinkNotSupported);
 
                 default:
                     Contract.Assert(false, "We should never get here, it means the ODataReader reported a wrong state.");
                     break;
             }
         }
+
+
+        private static void AddNestedResourceInfo(Stack<ODataItemBase> itemsStack, ODataResourceSetWrapperBase resourceSetWrapper)
+        {
+            ODataNestedResourceInfoWrapper parentNestedResourceInfo = (ODataNestedResourceInfoWrapper)itemsStack.Peek();
+            Contract.Assert(parentNestedResourceInfo != null, "this has to be an inner resource set. inner resource sets always have a nested resource info.");
```

Comment:
I doubt that we ever push a `null` into the stack. A `null` in the stack is the only thing that would give rise to this situation.
The only other scenario would be if `(ODataNestedResourceInfoWrapper)itemsStack.Peek()` was `itemsStack.Peek() as ODataNestedResourceInfoWrapper`